### DISCUSSION TITLE
Perspektive transformation

### DISCUSSION THREAD

**Discussion by Benik3** [libvips/libvips#3140] [Q&A]
Hello guys.
I would like to try to make perspective transformation using libvips.
I already did it using OpenCV, but because I need the transformation in "real time" (images from camera) I'm looking for something with better performance.

What I understand, I need to use image.mapim. But I have no idea how to calculate the matrix. Does anyone already tried this? I didn't found any code with this situation.

Thank you!

**Comment by jcupitt** [comment]
Hi @Benik3, I wrote an SO answer here:

https://stackoverflow.com/questions/50202733/how-to-perform-perspective-distort-transformation-in-vips

**Reply by jcupitt** [reply]
Though I don't know if performance will be better than OpenCV for this case -- transforming like this will not be super optimised, of course.

You could try rendering the transform at a lower resolution (x4 perhaps?) then scaling it up. That should get some extra speed and only lose a little accuracy.

**Comment by Benik3** [comment]
Thanks for the reply.
I already do it on 640x480 image. The transformation take around 7ms. With bigger image it go up pretty quickly and eat lot of CPU time.
I will try your example in C, but what you mean by comment "You'd also need something to calculate the transform from a set of tie-points, of course."?

**Reply by jcupitt** [reply]
That's the matrix T in that example.

There are lots of ways of calculating T -- I usually use least mean square from a set of tie points, but you can compute it eg. from a camera model too.

**Reply by Benik3** [reply]
Do you have please any link about calculation of the T matrix?

**Reply by Benik3** [reply]
I'm looking into OpenCV source code, and there is pretty nice comment.
https://github.com/opencv/opencv/blob/8f0edf6a1c7015dd76abb7dc8a82ea36d5500a72/modules/imgproc/src/imgwarp.cpp#L3383
If I understand it right, your mentioned T matrix is the coefficient matrix Cij in the comment right?
Now I just have to figure out how to calculate it :D

**Reply by jcupitt** [reply]
That's a good find. Yes, you'll need something like that.

**Comment by Benik3** [comment]
In the Ruby example there is `i = Vips::Image.xyz image.width, image.height`
Is it width and height of input or of output image?

**Reply by Benik3** [reply]
Anyway I get stuck on making of the x and y in CPP :/

**Reply by Benik3** [reply]
ok, it's just VImage object. I tried to use calculated matrix from OpenCV, but the resulted image is wrong.
Anyway it took much longer then in OpenCV (cca 30ms vs 7ms)

**Reply by jcupitt** [reply]
That seems a bit slow. Are you recomputing the transform every frame? You can reuse it, eg.:

```ruby
#!/usr/bin/ruby

require 'vips'

image_width = 640
image_height = 480

# perspective distortion: each pixel (x', y') in the output image is
# interpolated from pixel (x, y) in the input using:
#
#   x' = (A x + B y + C) / (G x + H y + 1)
#   y' = (D x + E y + F) / (G x + H y + 1)
#
# where the constants A .. H are from the transform matrix T
#
# T = [A, B, .. H]

T = [1.0, 0.0, 0.0, 0.0, 1.0, 0.0, 0.0003, 0.0001]

# make an index image where pixels have the value of their (x, y) coordinates
i = Vips::Image.xyz image_width, image_height

x = (i[0] * T[0] + i[1] * T[1] + T[2]) / (i[0] * T[6] + i[1] * T[7] + 1)
y = (i[0] * T[3] + i[1] * T[4] + T[5]) / (i[0] * T[6] + i[1] * T[7] + 1)

# join up x and y as a map image
m = x.bandjoin y

# and render to a memory image so we can reuse the result
m = m.copy_memory

total_time = 0
ARGV[0].to_i.times do
  # make an RGB memory image for us to transform (don't time this)
  frame = Vips::Image.black(image_width, image_height, bands: 3).copy_memory()
  
  # transform to another memory image (time this part)
  start = Time.now
  frame2 = frame.mapim(m).copy_memory
  stop = Time.now
  total_time += stop - start
end

puts "#{ARGV[0]} frames in #{total_time}s"
```

I see:

```
$ ./perspective.rb 1000
1000 frames in 3.1365763899999983s
```

So about 3ms per frame.

**Reply by jcupitt** [reply]
(though this is a pretty chunky PC)

**Reply by Benik3** [reply]
I'm running it on RPi4, but even on PC, If I remember, I didn't have better time with OpenCV.
Here is my usage (simplified):

```cpp
uint8_t *_outptr;
VImage _m;
// one time init in constructor
    VIPS_INIT("my-app");
    // _warp is matrix calculated by OpenCv
    double T[] = {_warp.ptr<double>()[0], _warp.ptr<double>()[1], _warp.ptr<double>()[2], _warp.ptr<double>()[3], _warp.ptr<double>()[4], _warp.ptr<double>()[5], _warp.ptr<double>()[6], _warp.ptr<double>()[7]}; 
    VImage i = VImage::xyz(640, 480);

    VImage xi = (i[0] * T[0] + i[1] * T[1] + T[2]) / (i[0] * T[6] + i[1] * T[7] + 1);
    VImage yi = (i[0] * T[3] + i[1] * T[4] + T[5]) / (i[0] * T[6] + i[1] * T[7] + 1);

    _m = xi.bandjoin(yi);

// this for each image data, img is pointer to source image array in memory
    VImage in = VImage::new_from_memory(img, 640*480*3, 640, 480, 3, VIPS_FORMAT_UCHAR);
    VImage out = in.mapim(_m);

    _outptr = (uint8_t *) out.write_to_memory(nullptr);

    return _outptr;

```
As input I have buffer of 4 images = pointers to memory in simple format RGBRGB.... This memory is allocated whole time of the program, so I just need to tell the VImage to take the data from another pointer. What I got from documentation, it's not supported this way, so I did the `new_from_memory`.  
I then need again pointer to the new image in the same form.  
In OpenCV I did it simply by:
```cpp
   // _src and _dts are global Mat objects
    _src->data = img;
    warpPerspective(*_src, *_dst, _warp, _dst->size(), INTER_NEAREST);
    return _dst->data;
```

**Reply by jcupitt** [reply]
That'll recompute the transform for every frame. Change it to:

```c++
    _m = xi.bandjoin(yi).copy_memory();
```

You should see a large speedup (a factor of five?).

OpenCV works in immediate mode, so when you call `warpPerspective` it executes that operation immediately.  By contrast, libvips is demand-driven, so when you call operations like `bandjoin`, you're not computing any pixels, you're just constructing a pipeline. Computation only happens when output pixels are generated, so in the final `write_to_memory()` in this case. 

`copy_memory()` will allocate ram for an image, then run prior pipeline stages into that. The `mapim` will then reuse the result and won't need to compute the whole pipeline. That's assuming you don't change the transform every frame, of course.

You can use nearest neighbor interpolation in ruby-vips with:

```ruby
nearest = Vips::Interpolate.new "nearest"
 ...

    frame2 = frame.mapim(m, interpolate: nearest).copy_memory
```

Which brings it down to 2.7ms per frame for me, so maybe 10% faster again.

**Reply by jcupitt** [reply]
... but you're right, this kind of task is exactly opencv's bread and butter and you're unlikely to be able to do better without a lot of work.

The rpi has quite a nice GPU, maybe running the transform in a shader would be the way to go? You could do better interpolation, and it'd certainly be quicker for screen output.

libvips can be faster than opencv for large images or complex pipelines, where chaining and threading become important.

**Reply by Benik3** [reply]
Thanks for the reply. I will test it tomorrow, when I will be back at home :)  
Yeah, using GPU would be probably much faster, but it's out of my skills :/  
I'm not showing the final picture on a screen, instead the data are sent through flatbuffer socket.

**Reply by Benik3** [reply]
I still got 30-40ms. Nevermind I will stick with OpenCV, thanks for your help!
